spec.html

```diff
@@ -7600,17 +7600,23 @@ <h1>GetGlobalObject ( )</h1>
   <emu-clause id="sec-jobs" oldids="sec-jobs-and-job-queues,sec-enqueuejob,sec-runjobs,job-queue">
     <h1>Jobs and Host Operations to Enqueue Jobs</h1>
     <p>A <dfn id="job">Job</dfn> is an abstract closure with no parameters that initiates an ECMAScript computation when no other ECMAScript computation is currently in progress.</p>
+    <p>A Job is stored in a <dfn>Job Record</dfn>, which has the slots [[AbstractClosure]], [[Realm]], and [[ScriptOrModule]].</p>
     <p>Jobs are scheduled for execution by ECMAScript host environments. This specification describes the host hook HostEnqueuePromiseJob to schedule one kind of job; host environments may define additional abstract operations which schedule jobs. Such operations accept a Job abstract closure as the parameter and schedule it to be performed at some future time. Their implementations must conform to the following requirements:</p>
 
     <ul>
-      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must:
-        <ol>
-          <li>Push an execution context onto the execution context stack.</li>
-          <li>Perform any implementation-defined preparation steps.</li>
-          <li>Call the abstract closure.</li>
-          <li>Perform any implementation-defined cleanup steps.</li>
-          <li>Pop the previously-pushed execution context from the execution context stack.</li>
-        </ol>
+      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must perform the following steps:
+        <emu-alg>
+          1. Let _job_ be the selected Job Record.
+          1. Let _context_ be a new ECMAScript code exceution context.
```

Comment:
If an execution context is not pushed, or one is pushed but the [[Realm]] is null, at least the following tests fail (i didn't bother with a full test262 run):

test/built-ins/Promise/create-resolving-functions-reject.js
test/built-ins/Promise/create-resolving-functions-resolve.js
test/built-ins/Promise/exception-after-resolve-in-executor.js
test/built-ins/Promise/exception-after-resolve-in-thenable-job.js
test/built-ins/Promise/resolve-prms-cstm-then-deferred.js
test/built-ins/Promise/resolve-prms-cstm-then-immed.js
test/built-ins/Promise/resolve-thenable-deferred.js
test/built-ins/Promise/resolve-thenable-immed.js
test/built-ins/Promise/all/reject-deferred.js
test/built-ins/Promise/all/reject-ignored-deferred.js
test/built-ins/Promise/all/reject-ignored-immed.js
test/built-ins/Promise/all/reject-immed.js
test/built-ins/Promise/all/resolve-thenable.js
test/built-ins/Promise/allSettled/reject-deferred.js
test/built-ins/Promise/allSettled/reject-ignored-immed.js
test/built-ins/Promise/allSettled/reject-ignored-deferred.js
test/built-ins/Promise/allSettled/reject-immed.js
test/built-ins/Promise/allSettled/resolve-thenable.js
test/built-ins/Promise/allSettled/resolved-then-catch-finally.js
test/built-ins/Promise/any/reject-deferred.js
test/built-ins/Promise/any/resolve-ignores-late-rejection.js
test/built-ins/Promise/any/resolve-ignores-late-rejection-deferred.js
test/built-ins/Promise/race/reject-deferred.js
test/built-ins/Promise/race/reject-ignored-deferred.js
test/built-ins/Promise/race/reject-ignored-immed.js
test/built-ins/Promise/race/reject-immed.js
test/built-ins/Promise/race/resolve-non-obj.js
test/built-ins/Promise/race/resolve-non-thenable.js
test/built-ins/Promise/race/resolve-poisoned-then.js
test/built-ins/Promise/race/resolve-prms-cstm-then.js
test/built-ins/Promise/race/resolve-thenable.js
test/built-ins/Promise/resolve/S25.4.4.5_A3.1_T1.js
test/built-ins/Promise/resolve/S25.Promise_resolve_foreign_thenable_1.js
test/built-ins/Promise/resolve/S25.Promise_resolve_foreign_thenable_2.js
test/built-ins/Promise/resolve/arg-uniq-ctor.js
test/built-ins/Promise/resolve/resolve-thenable.js
test/built-ins/Promise/prototype/finally/rejected-observable-then-calls-argument.js
test/built-ins/Promise/prototype/finally/rejected-observable-then-calls.js
test/built-ins/Promise/prototype/finally/rejection-reason-no-fulfill.js
test/built-ins/Promise/prototype/finally/resolution-value-no-override.js
test/built-ins/Promise/prototype/finally/resolved-observable-then-calls-argument.js
test/built-ins/Promise/prototype/finally/resolved-observable-then-calls.js
test/built-ins/Promise/prototype/finally/species-constructor.js
test/built-ins/Promise/prototype/finally/species-symbol.js
test/built-ins/Promise/prototype/finally/subclass-reject-count.js
test/built-ins/Promise/prototype/finally/subclass-resolve-count.js
test/built-ins/Promise/prototype/finally/subclass-species-constructor-reject-count.js
test/built-ins/Promise/prototype/finally/subclass-species-constructor-resolve-count.js
test/built-ins/Promise/prototype/then/resolve-pending-fulfilled-prms-cstm-then.js
test/built-ins/Promise/prototype/then/resolve-pending-fulfilled-thenable.js
test/built-ins/Promise/prototype/then/resolve-pending-rejected-prms-cstm-then.js
test/built-ins/Promise/prototype/then/resolve-pending-rejected-thenable.js
test/built-ins/Promise/prototype/then/resolve-settled-fulfilled-prms-cstm-then.js
test/built-ins/Promise/prototype/then/resolve-settled-fulfilled-thenable.js
test/built-ins/Promise/prototype/then/resolve-settled-rejected-prms-cstm-then.js
test/built-ins/Promise/prototype/then/resolve-settled-rejected-thenable.js
test/built-ins/WeakRef/prototype/deref/gc-cleanup-not-prevented-with-wr-deref.js
test/built-ins/FinalizationRegistry/gc-has-one-chance-to-call-cleanupCallback.js
test/built-ins/FinalizationRegistry/prototype/cleanupSome/cleanup-prevented-with-reference.js
test/built-ins/FinalizationRegistry/prototype/cleanupSome/cleanup-prevented-with-unregister.js
test/built-ins/FinalizationRegistry/prototype/cleanupSome/holdings-multiple-values.js
test/built-ins/FinalizationRegistry/prototype/cleanupSome/reentrancy.js
test/built-ins/FinalizationRegistry/prototype/cleanupSome/return-undefined-with-gc.js
test/built-ins/FinalizationRegistry/prototype/unregister/unregister-cleaned-up-cell.js

Reply:
Promises, weak refs, and finalizers are all features that are not otherwise observable except via host integration. So when testing those features, we must make _some_ assumptions about a test runner environment, and those assumptions are often stronger than the baseline normative requirements of ecma262.

Crucially, what you're pointing out is that the baseline normative requirements for job hooks, in this case, are insufficient if implemented as-is with no additional implementation-defined steps. But the job invariants are designed to give hosts freedom, since different jobs have different needs regarding execution contexts and realms. I've already pointed out a concrete case: PromiseReactionJob sometimes don't run user code at all because of revoked proxies, and so there is no reasonable Realm value to even pass.

I agree it'd be good to at least enumerate the invariants that would ensure asserts don't get tripped and don't leave that underspecified, but practically it must be done per-job hook.